solver/src/http_solver.rs

```diff
@@ -212,6 +224,43 @@ fn split_liquidity(liquidity: Vec<Liquidity>) -> (Vec<LimitOrder>, Vec<AmmOrder>
     (limit_orders, amm_orders)
 }
 
+fn remove_orders_without_native_connection(
+    orders: &mut Vec<LimitOrder>,
+    amms: &[AmmOrder],
+    native_token: &H160,
+) {
+    // Find all tokens that are connected through potentially multiple amm hops to the fee.
+    // TODO: Replace with a more optimal graph algorithm.
+    let mut amms = amms.iter().map(|amm| amm.tokens).collect::<HashSet<_>>();
+    let mut fee_connected_tokens = std::iter::once(*native_token).collect::<HashSet<_>>();
+    loop {
+        let mut added_token = false;
+        amms.retain(|token_pair| {
+            let tokens = token_pair.get();
+            if fee_connected_tokens.contains(&tokens.0) {
+                fee_connected_tokens.insert(tokens.1);
+                added_token = true;
+                false
+            } else if fee_connected_tokens.contains(&tokens.1) {
+                fee_connected_tokens.insert(tokens.0);
+                added_token = true;
+                false
+            } else {
+                true
+            }
+        });
+        if amms.is_empty() || !added_token {
```

Comment:
nanonit: curious why not a while loop? Not that it makes a big difference, I personally prefer while loops as they make the termination condition a bit more obvious than searching for the break.

Reply:
The is_empty could be moved to the while loop but the !added_token would still have to be checked inside and this way they are together. No preference either way.